lib/segment/src/index/field_index/geo_index.rs

```diff
@@ -1,180 +1,244 @@
-                .entry(geo_hash)
-                .or_insert_with(Vec::new)
-                .push(idx);
+                .entry(geo_hash.to_owned())
+                .or_insert_with(HashSet::new)
+                .insert(idx);
+
+            for i in 0..=geo_hash.len() {
+                let sub_geo_hash = &geo_hash[0..i];
+                match self.values_per_hash.get_mut(sub_geo_hash) {
+                    None => {
+                        self.values_per_hash.insert(sub_geo_hash.to_string(), 1);
+                    }
+                    Some(count) => {
+                        *count += 1;
+                    }
+                };
+                if !seen_hashes.contains(sub_geo_hash) {
+                    match self.points_per_hash.get_mut(sub_geo_hash) {
+                        None => {
+                            self.points_per_hash.insert(sub_geo_hash.to_string(), 1);
+                        }
+                        Some(count) => {
+                            *count += 1;
+                        }
+                    }
+                    seen_hashes.insert(sub_geo_hash);
+                }
+            }
         }
-        self.payload_count += values.len();
+
+        self.values_count += values.len();
         self.points_count += 1;
+        self.max_values_per_point = self.max_values_per_point.max(values.len());
+    }
+
+    fn get_stored_sub_regions(
+        &self,
+        geo: &GeoHash,
+    ) -> Box<dyn Iterator<Item = (&GeoHash, &HashSet<PointOffsetType>)> + '_> {
+        let geo_clone = geo.to_string();
+        Box::new(
+            self.points_map
+                .range(geo.to_string()..)
+                .take_while(move |(p, _h)| p.starts_with(&geo_clone)),
+        )
+    }
+
+    fn get_iterator(&self, values: Vec<GeoHash>) -> Box<dyn Iterator<Item = PointOffsetType> + '_> {
+        Box::new(
+            values
+                .into_iter()
+                .flat_map(|top_geo_hash| {
+                    self.get_stored_sub_regions(&top_geo_hash)
+                        .flat_map(|(_geohash, points)| points.iter().copied())
+                })
+                .unique(),
+        )
     }
 
-    fn get_iterator(&self, values: &[GeoHash]) -> Box<dyn Iterator<Item = PointOffsetType> + '_> {
-        let mut res: Vec<PointOffsetType> = vec![];
-        values.iter().for_each(|geo_hash| {
-            if let Some(vec) = self.points_map.get(geo_hash) {
-                res.extend(vec)
+    /// Get iterator over smallest geo-hash regions larger than `threshold` points
+    fn get_large_hashes(
+        &self,
+        threshold: usize,
+    ) -> Box<dyn Iterator<Item = (&GeoHash, usize)> + '_> {
+        let mut large_regions = self
+            .points_per_hash
+            .iter()
+            .filter(|(hash, size)| **size > threshold && !hash.is_empty())
+            .collect_vec();
+
+        // smallest regions first
+        large_regions.sort();
+        large_regions.reverse();
```

Comment:
I guess we could also sort it the other way around directly.

`large_regions.sort_by(|a, b| b.cmp(a));`

Reply:
:+1: